README.md
# (MODX)EvolutionCMS.snippets.ddIf

This snippet compares different values and returns required chunk or string.


## Requires

* PHP >= 5.6
* [(MODX)EvolutionCMS](https://github.com/evolution-cms/evolution) >= 1.1
* [(MODX)EvolutionCMS.libraries.ddTools](https://code.divandesign.ru/modx/ddtools) >= 0.60


## Installation


### Using [(MODX)EvolutionCMS.libraries.ddInstaller](https://github.com/DivanDesign/EvolutionCMS.libraries.ddInstaller)

Just run the following PHP code in your sources or [Console](https://github.com/vanchelo/MODX-Evolution-Ajax-Console):

```php
//Include (MODX)EvolutionCMS.libraries.ddInstaller
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddInstaller/require.php'
);

//Install (MODX)EvolutionCMS.snippets.ddIf
\DDInstaller::install([
	'url' => 'https://github.com/DivanDesign/EvolutionCMS.snippets.ddIf',
	'type' => 'snippet'
]);
```

* If `ddIf` is not exist on your site, `ddInstaller` will just install it.
* If `ddIf` is already exist on your site, `ddInstaller` will check it version and update it if needed.


### Manually


#### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddIf`.
2. Description: `<b>2.3.1</b> This snippet compares different values and returns required chunk or string.`.
3. Category: `Core`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddIf_snippet.php` file from the archive.


#### 2. Elements → Manage Files

1. Create a new folder `assets/snippets/ddIf/`.
2. Extract the archive to the folder (except `ddIf_snippet.php`).


## Parameters description

* `operand1`
	* Desctription: The first operand for comparing.  
		An empty unparsed placeholder (like `'[+somePlaceholder+]'`) will be interpretated as empty string (`''`).
	* Valid values: `string`
	* **Required**
	
* `operand2`
	* Desctription: The second operand for comparing.
	* Valid values: `string`
	* Default value: `''`
	
* `operator`
	* Desctription: Comparing operator.  
		Values are case insensitive (the following names are equal: `'isNumeric'`, `'isnumeric'`, `'ISNUMERIC'`, etc).
	* Valid values:
		* `'=='`
		* `'!='`
		* `'>'`
		* `'<'`
		* `'<='`
		* `'>='`
		* `'bool'`
		* `'inArray'`
		* `'isNumeric'`
		* `'isWhitespace'` — checks if `operand1` is just white space (an empty string is also considered as white space)
		* `'isIncludes'` — case-sensitive check if `operand1` contains `operand2`
	* Default value: `'=='`
	
* `trueChunk`
	* Desctription: This value is returning if result is true.
		Available placeholders:
		* `[+ddIfParams.operand1+]` — contains `operand1` value.
		* `[+ddIfParams.operand2+]` — contains `operand2` value.
		* `[+ddIfParams.operator+]` — contains `operator` value.
		* `[+`any placeholders from the `placeholders` parameter`+]`
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `''`
	
* `falseChunk`
	* Desctription: This value is returning if result is false. 
		Available placeholders:
		* `[+ddIfParams.operand1+]` — contains `operand1` value.
		* `[+ddIfParams.operand2+]` — contains `operand2` value.
		* `[+ddIfParams.operator+]` — contains `operator` value.
		* `[+`any placeholders from the `placeholders` parameter`+]`
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `''`
	
* `placeholders`
	* Desctription:
		Additional data has to be passed into the `trueChunk` and `falseChunk`.  
		Nested objects and arrays are supported too:
		* `{"someOne": "1", "someTwo": "test" }` => `[+someOne+], [+someTwo+]`.
		* `{"some": {"a": "one", "b": "two"} }` => `[+some.a+]`, `[+some.b+]`.
		* `{"some": ["one", "two"] }` => `[+some.0+]`, `[+some.1+]`.
	* Valid values:
		* `stringJsonObject` — as [JSON](https://en.wikipedia.org/wiki/JSON)
		* `stringHjsonObject` — as [HJSON](https://hjson.github.io/)
		* `stringQueryFormatted` — as [Query string](https://en.wikipedia.org/wiki/Query_string)
		* It can also be set as a native PHP object or array (e. g. for calls through `$modx->runSnippet`):
			* `arrayAssociative`
			* `object`
	* Default value: —
	
* `debugTitle`
	* Desctription: The title for the System Event log if debugging is needed.  
		Just set it and watch the System Event log.
	* Valid values: `string`
	* Default value: —


## Examples


### String comparison

```
[[ddIf?
	&operand1=`Test string 1`
	&operator=`==`
	&operand2=`Test string 2`
	&trueChunk=`@CODE:The strings are equal.`
	&falseChunk=`@CODE:The strings are not equal.`
]]
```

Returns:

```
The strings are not equal.
```


### Checks if a value exists in an array

```
[[ddIf?
	&operand1=`Apple`
	&operator=`inArray`
	&operand2=`Pear,Banana,Apple,Orange`
	&trueChunk=`@CODE:Exists.`
	&falseChunk=`@CODE:Not exists.`
]]
```

Returns:

```
Exists.
```


### Check if `operand1` contains `operand2`

```
[[ddIf?
	&operand1=`The quick brown fox jumps over the lazy dog.`
	&operator=`isIncludes`
	&operand2=`fox`
	&trueChunk=`@CODE:“fox” is found`
	&falseChunk=`@CODE:“fox” is not found`
]]
```

Returns:

```
“fox” is found
```


### Checks if a `operand1` value is number or not

```
[[ddIf?
	&operand1=`123`
	&operator=`isNumeric`
	&trueChunk=`@CODE:Number.`
	&falseChunk=`@CODE:Not number.`
]]
```

Returns:

```
Number.
```


### Checks if a `operand1` value is just white space or not

Any number of spaces / tabs / new lines / etc are considered as white space.
An empty string is also considered as white space.

```
[[ddIf?
	&operand1=`
		 
	   
	    
	`
	&operator=`isWhitespace`
	&trueChunk=`@CODE:The string contains only some whitespace characters.`
	&falseChunk=`@CODE:[+ddIfParams.operand1+]`
]]
```

Returns:

```
The string contains only some whitespace characters.
```

If `operand1` contains any non-whitespace characters, `falseString` will be returned.

```
[[ddIf?
	&operand1=`All you need is love.`
	&operator=`isWhitespace`
	&trueChunk=`@CODE:The string contains only some whitespace characters.`
	&falseChunk=`@CODE:[+ddIfParams.operand1+]`
]]
```

Returns:

```
All you need is love.
```


### Number comparison and pass additional data to chunks

```
[[ddIf?
	&operand1=`[*general_price*]`
	&operator=`<`
	&operand2=`500`
	&trueChunk=`general_goodInexpensive`
	&falseChunk=`general_good`
	&placeholders=`{
		"title": "[*pagetitle*]",
		"image": "[*general_image*]",
		"somethingText": "All we need is love!"
	}`
]]
```

Code of the `general_good` chunk:

```html
<div>
	<h2>[+title+], $[+ddIfParams.operand1+]</h2>
	<img src="[+image+]" alt="[+title+]" />
</div>
```

Code of the `general_goodInexpensive` chunk:

```html
<div class="inexpensive">
	<h2>[+title+], <strong>$[+ddIfParams.operand1+]</strong></h2>
	<img src="[+image+]" alt="[+title+]" />
	<p>[+somethingText+]</p>
</div>
```

Let `[*general_price*]` be equal to `120`, then the snippet returns:

```html
<div class="inexpensive">
	<h2>Some inexpensive good, <strong>$120</strong></h2>
	<img src="assets/images/goods/some1.jpg" alt="Some inexpensive good" />
</div>
```


### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
//Include (MODX)EvolutionCMS.libraries.ddTools
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddTools/modx.ddtools.class.php'
);

//Run (MODX)EvolutionCMS.snippets.ddIf
\DDTools\Snippet::runSnippet([
	'name' => 'ddIf',
	'params' => [
		'operand1' => '1',
		'operator' => 'bool',
		'trueChunk' => '@CODE:It's true!'
	]
]);
```


## Links

* [Home page](https://code.divandesign.ru/modx/ddif)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddif)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.snippets.ddIf)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddIf changelog


## Version 2.3.1 (2023-06-03)
* \* Compatibility with PHP < 8.
* \* Parameters → `trueChunk`, `falseChunk` → Placeholders: The following have been renamed (outdated names are still supported, but are not recommended):
	* \* `[+snippetParams.operand1+]` → `[+ddIfParams.operand1+]`.
	* \* `[+snippetParams.operand2+]` → `[+ddIfParams.operand2+]`.
	* \* `[+snippetParams.operator+]` → `[+ddIfParams.operator+]`.


## Version 2.3 (2023-06-03)
* \+ Parameters → `operator` → Valid values → `'isIncludes'`: The new valid value. Case-sensitive check if `operand1` contains `operand2` (see README → Examples).
* \* `\ddTools::getTpl` is used instead of `$modx->getTpl` (means a bit less bugs).
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.60 is required.


## Version 2.2 (2021-04-30)
* \+ Parameters → `operator`:
	* \+ Added the ability to check if `operand1` is just white space (see README).
	* \+ Values are case insensitive.
* \+ README → Documentation → Installation → Using (MODX)EvolutionCMS.libraries.ddInstaller.


## Version 2.1 (2021-04-27)
* \* Attention! PHP >= 5.6 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.49.1 is required.
* \+ Parameters → `placeholders`: Can also be set as [HJSON](https://hjson.github.io/).
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval (see README → Examples).
* \+ `\ddIf\Snippet`: The new class. All snippet code was moved here.
* \+ Composer.json → `support`.


## Version 2.0 (2021-01-20)
* \* Attention! Backward compatibility is broken.
* \- Parameters → `trueString`, `falseString`: The deprecated parameters are no longer supported.
* \* Parameters → `trueChunk`, `falseChunk`: The following available placeholders were renamed:
	* \* `ddIf_operand1` → `snippetParams.operand1`.
	* \* `ddIf_operand2` → `snippetParams.operand2`.
	* \* `ddIf_operator` → `snippetParams.operator`.
* \+ Snippet result was added to debug log message.
* \+ Improved style of debug log message.
* \* Refactoring:
	* \* The `$params` variable is used instead of a standalone variable for each parameter.
	* \* `\DDTools\ObjectTools::extend` is used to set parameters defaults.
* \+ README → Links → Packagist.


## Version 1.7.1 (2020-10-08)
* \* Parameters → `operand1`: Improved an empty unparsed placeholder recognition.


## Version 1.7 (2020-09-28)
* \* Attention! (MODX)Evolution.libraries.ddTools >= 0.40.1 is required (not tested with older versions).
* \+ Parameters → `placeholders`: Nested arrays are supported too.
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.
* \+ Composer.json.


## Version 1.6 (2018-12-05)
* \+ Parameters → `trueChunk`, `falseChunk`: Added few placeholders:
	* \+ `[+ddIf_operand1+]` — contains `operand1` value.
	* \+ `[+ddIf_operand2+]` — contains `operand2` value.
	* \+ `[+ddIf_operator+]` — contains `operator` value.


## Version 1.5 (2018-10-31)
* \+ Parameters → `debugTitle`: The new parameter. If you need to debug just set it and watch the System Event log.
* \+ Parameters → `operand1`: If is equal to something like `[+somePlaceholder+]` the snippet interpretate it as `''`. It's convenient when placeholders were not parsed for any reason.


## Version 1.4 (2017-02-10)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)Evolution >= 1.1 is required.
* \* Attention! (MODX)Evolution.libraries.ddTools >= 0.18 is required.
* \+ Parameters → `placeholders`: Added JSON and Query string support. The old format is still supported but deprecated.
* \+ Parameters → `trueChunk`, `falseChunk`: Added support of the `@CODE:` keyword prefix.
* \- Parameters → `trueString`, `falseString`: Are deprecated (use `@CODE:` prefix instead). Backward compatibility is preserved with message to the CMS Event log.
* \* Small optimization and other changes.


## Version 1.3 (2016-07-14)
* \+ Parameter → `operator`:
	* \+ Now can be equal to `isnumeric` — check the `operand1` is number or not.
	* \+ Is not case sensitive.


## Version 1.2 (2015-02-22)
* \* Attention! (MODX)Evolution >= 1.0.13 is required.
* \+ Parameters → `operator`: Now can gets a simple operator value instead of it’s letter alias (of course, with backward compatibility).
* \+ Additional data (from the `placeholders` parameter) are processed in the `trueString` and `falseString` parameters too.
* \* Refactoring:
	* \* The `$result` variable has not been used for placeholders array anymore, the `$placeholder` variable instead.
	* \* One `return` instead of several, `return` in any case.


## Version 1.1 (2012-04-05)
* \* Attention! Backward compatibility is broken. Actually the version should be named as 2.0 because of backward compatibility absence but renaming hasn’t been done on account of minor changes.
* \* Attention! (MODX)Evolution.libraries.ddTools >= 0.2 is required.
* \* Parameters:
	* \* `operand1`: Has been renamed from `subject`.
	* \* `operand2`: Has been renamed from `operand`.
	* \* `trueChunk`: Has been renamed from `trueChank`.
	* \* `falseChunk`: Has been renamed from `falseChank`.


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />
<style>ul{list-style:none;}</style>

ddIf.php
<?php
/**
 * ddIf
 * @version 1.5 (2018-10-31)
 * 
 * @desc This snippet compares 2 values and returns required chunk or string.
 * 
 * @uses PHP >= 5.4.
 * @uses MODXEvo >= 1.1.
 * @uses MODXEvo.library.ddTools >= 0.18.
 * 
 * @param $operand1 {string} — The first operand for comparing. @required
 * @param $operand2 {string} — The second operand for comparing. Default: ''.
 * @param $operator {'=='|'!='|'>'|'<'|'<='|'>='|'bool'|'inarray'|'isnumeric'} — Comparing operator. Default: '=='.
 * @param $trueChunk {string_chunkName|string} — This value is returning if result is true (chunk name or code via “@CODE:” prefix). Default: ''.
 * @param $falseChunk {string_chunkName|string} — This value is returning if result is false (chunk name or code via “@CODE:” prefix). Default: ''.
 * @param $placeholders {string_json|string_queryFormated} — Additional data which is required to transfer to chunk. JSON or query-formated string, e. g.: '{"width": 800, "height": 600}' or 'width=800&height=600'. Default: ''.
 * @param $debugTitle {string} — The title for the System Event log if debugging is needed.
 * 
 * @link http://code.divandesign.biz/modx/ddif/1.5
 * 
 * @copyright 2012–2018 DivanDesign {@link http://www.DivanDesign.biz }
 */

$result = '';

require_once $modx->getConfig('base_path').'assets/libs/ddTools/modx.ddtools.class.php';

//Если для отладки нужно вывести то что пришло в сниппет выводим
if(isset($debugTitle)){
	ddTools::logEvent([
		'message' => '<p>Snippet parameters:</p><code><pre>'.var_export(
			$params,
			true
		).'</pre></code>',
		'source' => 'ddIf: '.$debugTitle
	]);
}

//Если передано, что сравнивать
if (isset($operand1)){
	//Если это сырой плейсхолдер, то скорее всего он пустой, и его не обработал парсер, приравняем тогда параметр к пустоте
	if(mb_substr(
		$operand1,
		0,
		2
	) == '[+'){
		$operand1 = '';
	}
	
	//Если передали, с чем сравнивать, хорошо, если нет — будем с пустой строкой
	$operand2 = isset($operand2) ? $operand2 : '';
	$operator = isset($operator) ? mb_strtolower($operator) : '==';
	
	//Булевое значение истинности сравнения
	$boolOut = '';
	//Выбираем сравнение в зависимости от оператора
	switch ($operator){
		case '!=':
		//Backward compatibility
		case '!r':
			$boolOut = ($operand1 != $operand2) ? true : false;
		break;
		
		case '>':
		case 'b':
			$boolOut = ($operand1 > $operand2) ? true : false;
		break;
		
		case '<':
		case 'm':
			$boolOut = ($operand1 < $operand2) ? true : false;
		break;
		
		case '>=':
		case 'br':
			$boolOut = ($operand1 >= $operand2) ? true : false;
		break;
		
		case '<=':
		case 'mr':
			$boolOut = ($operand1 <= $operand2) ? true : false;
		break;
		
		case 'bool':
			$boolOut = ($operand1) ? true : false;
		break;
		
		case 'inarray':
			$operand2 = explode(
				',',
				$operand2
			);
			$boolOut = in_array(
				$operand1,
				$operand2
			) ? true : false;
		break;
		
		case 'isnumeric':
			$boolOut = is_numeric($operand1);
		break;
		
		case '==':
		case 'r':
		default:
			$boolOut = ($operand1 == $operand2) ? true : false;
	}
	
	//Если есть дополнительные данные
	if (isset($placeholders)){
		//Разбиваем их
		$placeholders = ddTools::encodedStringToArray($placeholders);
	}else{
		$placeholders = [];
	}
	
	//Backward compatibility
	if (
		isset($trueString) ||
		isset($falseString)
	){
		ddTools::logEvent([
			'message' => '<p>The “trueString” and “falseString” parameters are deprecated. Please use instead “trueChunk” and “falseChunk” with the “@CODE:” prefix.</p>'
		]);
		
		if (isset($trueString)){$trueChunk = '@CODE:'.$trueString;}
		if (isset($falseString)){$falseChunk = '@CODE:'.$falseString;}
	}
	
	//Select output chunk
	$resultChunk = $boolOut ? (isset($trueChunk) ? $trueChunk : '') : (isset($falseChunk) ? $falseChunk : '');
	
	//$modx->getTpl('@CODE:') returns '@CODE:' O_o
	$resultChunk = $modx->getTpl($resultChunk != '@CODE:' ? $resultChunk : '');
	
	$result = $modx->parseText(
		$resultChunk,
		$placeholders
	);
}

return $result;
?>

src/Snippet.php
<?php
namespace ddIf;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '2.3.1',
		
		$params = [
			//Defaults
			//Required
			'operand1' => null,
			//Если передали, с чем сравнивать, хорошо, если нет — будем с пустой строкой
			'operand2' => '',
			'operator' => '==',
			'trueChunk' => '',
			'falseChunk' => '',
			'placeholders' => [],
			//Unset
			'debugTitle' => null
		],
		
		$paramsTypes = [
			'placeholders' => 'objectArray'
		]
	;
	
	/**
	 * prepareParams
	 * @version 1.2 (2021-04-30)
	 * 
	 * @param $params {stdClass|arrayAssociative|stringJsonObject|stringQueryFormatted}
	 * 
	 * @return {void}
	 */
	protected function prepareParams($params = []){
		//Call base method
		parent::prepareParams($params);
		
		//Если это сырой плейсхолдер, то скорее всего он пустой, и его не обработал парсер, приравняем тогда параметр к пустоте
		if (
			is_string($this->params->operand1) &&
			mb_substr(
				$this->params->operand1,
				0,
				2
			) == '[+' &&
			mb_substr(
				$this->params->operand1,
				-2
			) == '+]'
		){
			$this->params->operand1 = '';
		}
		
		$this->params->operator = mb_strtolower($this->params->operator);
		
		//Backward compatibility
		$operatorBackwardCompliance = [
			'r' => '==',
			'!r' => '!=',
			'b' => '>',
			'm' => '<',
			'br' => '>=',
			'mr' => '<='
		];
		
		if (
			array_key_exists(
				$this->params->operator,
				$operatorBackwardCompliance
			)
		){
			$this->params->operator = $operatorBackwardCompliance[$this->params->operator];
		}
	}
	
	/**
	 * run
	 * @version 1.2.3 (2023-06-03)
	 * 
	 * @return {string}
	 */
	public function run(){
		//The snippet must return an empty string even if result is absent
		$result = '';
		
		//Если передано, что сравнивать
		if (!is_null($this->params->operand1)){
			//Булевое значение истинности сравнения
			$boolOut = '';
			
			//Выбираем сравнение в зависимости от оператора
			switch ($this->params->operator){
				case '!=':
					$boolOut = $this->params->operand1 != $this->params->operand2;
				break;
				
				case '>':
					$boolOut = $this->params->operand1 > $this->params->operand2;
				break;
				
				case '<':
					$boolOut = $this->params->operand1 < $this->params->operand2;
				break;
				
				case '>=':
					$boolOut = $this->params->operand1 >= $this->params->operand2;
				break;
				
				case '<=':
					$boolOut = $this->params->operand1 <= $this->params->operand2;
				break;
				
				case 'bool':
					$boolOut = boolval($this->params->operand1);
				break;
				
				case 'isincludes':
					$boolOut =
						function_exists('str_contains') ?
						//PHP >= 8
						str_contains(
							$this->params->operand1,
							$this->params->operand2
						) :
						//PHP < 8
						(
							$this->params->operand2 === '' ||
							mb_strpos(
								$this->params->operand1,
								$this->params->operand2
							) !== false
						)
					;
				break;
				
				case 'inarray':
					$operand2Array = explode(
						',',
						$this->params->operand2
					);
					
					$boolOut = in_array(
						$this->params->operand1,
						$operand2Array
					);
				break;
				
				case 'isnumeric':
					$boolOut = is_numeric($this->params->operand1);
				break;
				
				case 'iswhitespace':
					$boolOut =
						$this->params->operand1 == '' ||
						ctype_space($this->params->operand1)
					;
				break;
				
				case '==':
				default:
					$boolOut = $this->params->operand1 == $this->params->operand2;
			}
			
			//Select output chunk
			$resultChunk =
				$boolOut ?
				$this->params->trueChunk :
				$this->params->falseChunk
			;
			
			$resultChunk = \ddTools::getTpl($resultChunk);
			
			$result = \ddTools::parseText([
				'text' => $resultChunk,
				'data' => \DDTools\ObjectTools::extend([
					'objects' => [
						[
							'ddIfParams.operand1' => $this->params->operand1,
							'ddIfParams.operand2' => $this->params->operand2,
							'ddIfParams.operator' => $this->params->operator,
							//Backward compatibility
							'snippetParams.operand1' => $this->params->operand1,
							'snippetParams.operand2' => $this->params->operand2,
							'snippetParams.operator' => $this->params->operator,
						],
						$this->params->placeholders
					]
				])
			]);
		}
		
		//Если для отладки нужно вывести то что пришло в сниппет выводим
		if(!is_null($this->params->debugTitle)){
			\ddTools::logEvent([
				'message' =>
					'<p>Snippet parameters:</p><pre><code>' .
					var_export(
						$this->params,
						true
					) .
					'</code></pre>' .
					'<p>Snippet result:</p><pre><code>' .
					var_export(
						$result,
						true
					) .
					'</code></pre>'
				,
				'source' =>
					'ddIf: ' .
					$this->params->debugTitle
			]);
		}
		
		return $result;
	}
}